specs/BlockchainLayer/EpochManager/Epoch.md

```diff
@@ -0,0 +1,41 @@
+# Epoch
+
+All blocks are split into epochs. Within one epoch, the set of validators is fixed, and validator rotation
+happens at epoch boundaries.
+
+Genesis block is in its own epoch. After that, a block is either in its parent's epoch or
+starts a new epoch if it meets certain conditions.
+
+### BlockHeight
+Each block has a height. Genesis block has height 0, the height of each block is greater than
+the height of its parent. Within one epoch, each height has a block producer assigned to it.
+
+### End of an epoch
+A block is defined to be the last block in its epoch if it's the genesis block or if the following conditions are met:
+- Let `estimated_next_epoch_start = first_block_in_epoch.height + epoch_length`
+- `block.height + 1 >= estimated_next_epoch_start`
+- `block.last_finalized_height + 3 >= estimated_next_epoch_start`
+
+`epoch_length` is defined in `genesis_config` and has a value of `43200` blocks on mainnet (12 hours at 1 block per second).
+
+### EpochHeight
+Epochs on one chain can be identified by height, which is defined the following way:
+- Special epoch that contains genesis block only: undefined
+- Epoch starting from the block that's after genesis: `0` (NOTE: in current implementation it's `1` due to a bug, so there are two epochs with height `1`)
```

Comment:
I think the height of this epoch is intended to be 1 and the genesis epoch is intended to have height 0

Reply:
We should fix the code - either shift all epochs by one or change this one to 0 (and the first one can stay undefined - it doesn't have EpochInfo). Shifting all epochs might break explorers/stake monitoring tools?

Reply:
Agree. It doesn't seem to me that it would break anything if we change the epoch height of genesis to 0 and keep all other epoch heights the same. @SkidanovAlex does bridge use epoch heights at all?